modules/phase_field/src/postprocessors/TimeSeriesAverage.C

```diff
@@ -0,0 +1,98 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "TimeSeriesAverage.h"
+#include "FEProblem.h"
+
+template <>
+InputParameters
+validParams<TimeSeriesAverage>()
+{
+  InputParameters params = validParams<GeneralPostprocessor>();
+  params.addRequiredParam<PostprocessorName>("average", "The postprocessor to average");
+  params.addParam<unsigned int>("steps", "Number of previous steps to average over");
+  params.addParam<Real>("time", "Time interval to average over");
+  return params;
+}
+
+TimeSeriesAverage::TimeSeriesAverage(const InputParameters & parameters)
+  : GeneralPostprocessor(parameters),
+    _feproblem(dynamic_cast<FEProblemBase &>(_subproblem)),
+    _slave_pp(getPostprocessorValue("average")),
+    _average_mode(isParamValid("steps") ? AverageMode::STEPS : AverageMode::TIME),
+    _previous_steps(isParamValid("steps") ? getParam<unsigned int>("steps") : 0),
+    _time_window(isParamValid("time") ? getParam<Real>("time") : 0.0),
+    _history(0)
+{
+  if (isParamValid("steps") == isParamValid("time"))
+    mooseError("Specify exactly one of the 'steps' and 'time' parameters");
+}
+
+void
+TimeSeriesAverage::execute()
+{
+  // push current state into history
+  HistoryItem current;
+  current._dt = _feproblem.dt();
+  current._value = _slave_pp;
+  _history.push_back(current);
+}
+
+Real
+TimeSeriesAverage::getValue()
+{
+  // At the first time step we return 0.0, otherwise perform desired average
+  if (_history.size() == 0)
+    return 0.0;
+
+  switch (_average_mode)
+  {
+    case AverageMode::STEPS:
+      return stepAverage();
+
+    case AverageMode::TIME:
+      return timeAverage();
+  }
+
+  mooseError("Invalid average mode.");
+}
+
+Real
+TimeSeriesAverage::stepAverage() const
+{
+  int i = _history.size();
```

Comment:
Are you using signed values here to handle the empty case? You might need an explicit cast to avoid compiler warnings if that's what you are after.